pkg/server/endpoints/registration/handler.go

```diff
@@ -545,6 +547,95 @@ func (h *Handler) ListAgents(ctx context.Context, listReq *registration.ListAgen
 	return &registration.ListAgentsResponse{Nodes: resp.Nodes}, nil
 }
 
+func (h *Handler) MintX509SVID(ctx context.Context, req *registration.MintX509SVIDRequest) (_ *registration.MintX509SVIDResponse, err error) {
+	counter := telemetry_registrationapi.StartMintX509SVIDCall(h.Metrics)
+	addCallerIDLabel(ctx, counter)
+	defer counter.Done(&err)
+
+	spiffeID, err := h.normalizeSPIFFEIDForMinting(req.SpiffeId)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(req.Csr) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "request missing CSR")
+	}
+
+	for _, dnsName := range req.DnsNames {
+		if err := validateDNS(dnsName); err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "%q is not a valid DNS name: %v", dnsName, err)
+		}
+	}
+
+	csr, err := x509.ParseCertificateRequest(req.Csr)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: %v", err)
+	}
+	if err := csr.CheckSignature(); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid CSR: signature verify failed")
+	}
+
+	svid, err := h.ServerCA.SignX509SVID(ctx, ca.X509SVIDParams{
+		SpiffeID:  spiffeID,
+		PublicKey: csr.PublicKey,
+		TTL:       time.Duration(req.Ttl) * time.Second,
+		DNSList:   req.DnsNames,
+	})
+
+	resp, err := h.getDataStore().FetchBundle(ctx, &datastore.FetchBundleRequest{
+		TrustDomainId: h.TrustDomain.String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.Bundle == nil {
+		return nil, errors.New("bundle not found")
```

Comment:
We should probably return grpc error codes for these two conditions. Perhaps internal and failed_precondition respectively?

Reply:
Good call. Will fix.